### ISSUE TITLE
MySQL 8 with mariadb-connector-c 3.0.8-r0: Plugin caching_sha2_password could not be loaded

### ISSUE THREAD

**Issue by harmdewit** [brianmario/mysql2#1023]
My setup:

- Server running MySQL 8.0.14
- Alpine 3.9 (Docker)
- Ruby 2.4.5
- mariadb-connector-c-dev 3.0.8-r0 (this version added caching_sha2_password support)
- mysql2-0.5.2

When trying to run the application I get this error:

```
setup  | /usr/local/bundle/gems/mysql2-0.5.2/lib/mysql2/client.rb:90:in `connect': Plugin caching_sha2_password could not be loaded: Error loading shared library lib/mariadb/plugin/caching_sha2_password.so: No such file or directory (Mysql2::Error)
setup  | 	from /usr/local/bundle/gems/mysql2-0.5.2/lib/mysql2/client.rb:90:in `initialize'
setup  | 	from /usr/local/bundle/gems/activerecord-4.2.11/lib/active_record/connection_adapters/mysql2_adapter.rb:18:in `new'
setup  | 	from /usr/local/bundle/gems/activerecord-4.2.11/lib/active_record/connection_adapters/mysql2_adapter.rb:18:in `mysql2_connection'
setup  | 	from /usr/local/bundle/gems/activerecord-4.2.11/lib/active_record/connection_adapters/abstract/connection_pool.rb:438:in `new_connection'
setup  | 	from /usr/local/bundle/gems/activerecord-4.2.11/lib/active_record/connection_adapters/abstract/connection_pool.rb:448:in `checkout_new_connection'
setup  | 	from /usr/local/bundle/gems/activerecord-4.2.11/lib/active_record/connection_adapters/abstract/connection_pool.rb:422:in `acquire_connection'
setup  | 	from /usr/local/bundle/gems/activerecord-4.2.11/lib/active_record/connection_adapters/abstract/connection_pool.rb:349:in `block in checkout'
setup  | 	from /usr/local/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
setup  | 	from /usr/local/bundle/gems/activerecord-4.2.11/lib/active_record/connection_adapters/abstract/connection_pool.rb:348:in `checkout'
setup  | 	from /usr/local/bundle/gems/activerecord-4.2.11/lib/active_record/connection_adapters/abstract/connection_pool.rb:263:in `block in connection'
setup  | 	from /usr/local/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
setup  | 	from /usr/local/bundle/gems/activerecord-4.2.11/lib/active_record/connection_adapters/abstract/connection_pool.rb:262:in `connection'
setup  | 	from /usr/local/bundle/gems/activerecord-4.2.11/lib/active_record/connection_adapters/abstract/connection_pool.rb:571:in `retrieve_connection'
setup  | 	from /usr/local/bundle/gems/activerecord-4.2.11/lib/active_record/connection_handling.rb:113:in `retrieve_connection'
setup  | 	from /usr/local/bundle/gems/activerecord-4.2.11/lib/active_record/connection_handling.rb:87:in `connection'
setup  | 	from /usr/local/bundle/gems/activerecord-4.2.11/lib/active_record/migration.rb:391:in `check_pending!'
```

Running `mariadb_config` (or `mysql_config`) returns:

```bash
/code # mysql_config
Copyright 2011-2015 MariaDB Corporation AB
Get compiler flags for using the MariaDB Connector/C.
Usage: mysql_config [OPTIONS]
  --cflags        [-I/usr/include/mysql -I/usr/include/mysql]
  --include       [-I/usr/include/mysql -I/usr/include/mysql]
  --libs          [-L/usr/lib/ -lmariadb -lz -lssl -lcrypto]
  --libs_r        [-L/usr/lib/ -lmariadb -lz -lssl -lcrypto]
  --libs_sys      [-lz -lssl -lcrypto]
  --version       [10.3.6]
  --cc_version    [3.0.8]
  --socket        [/run/mysqld/mysqld.sock]
  --port          [3306]
  --plugindir     [/usr/lib/mariadb/plugin]
  --tlsinfo       [OpenSSL 1.1.1a]

# Listing plugins
/code # ls /usr/lib/mariadb/plugin
caching_sha2_password.so  dialog.so                 mysql_clear_password.so   sha256_password.so


# The mysql client does work correctly
/code # mysql -h host.docker.internal
Welcome to the MariaDB monitor.  Commands end with ; or \g.
```

It seems like mysql2 tries to load the plugins from a relative path, i.e. `./lib/mariadb/plugin/caching_sha2_password.so`. When I symlink the directory relative to the application then it works:

```
/code # bundle exec rails r "puts 'success'"
/usr/local/bundle/gems/mysql2-0.5.2/lib/mysql2/client.rb:90:in `connect': Plugin caching_sha2_password could not be loaded: Error loading shared library lib/mariadb/plugin/caching_sha2_password.so: No such file or directory (Mysql2::Error)

/code # mkdir ./lib/mariadb && ln -s /usr/lib/mariadb/plugin ./lib/mariadb/plugin
/code # bundle exec rails r "puts 'success'"
success
```

Could it be that `mysql2` ignores the plugindir from the mysql configuration? I see several refences to `--include` and --libs`  in the code below but nothing with `--plugin`:

https://github.com/brianmario/mysql2/blob/98304cbbc0c9a964c833474a06f55c433ec26432/ext/mysql2/extconf.rb#L75-L83

**Comment by TuxmAL** [comment]
Same issue here with:
- Server (docker contianer)
  - Debian 9 (stretch);
  - MySQL community server ver. 8.0.13, Protocol ver. 10.
- Client:
  - Fedora 29 (KDE spin);
  - Ruby ver. 2.4.5;
  - mysql2 ver. 0.5.2;
  - mariadb-connector-c.x86_64.3.0.8-1;
  - mariadb-connector-c-devel.x86_64-3.0.8-1;
  - mariadb-connector-c-config.noarch.3.0.8-1;

Running `mariadb_config` it reports:

```
Copyright 2011-2015 MariaDB Corporation AB
Get compiler flags for using the MariaDB Connector/C.
Usage: mariadb_config [OPTIONS]
  --cflags        [-I/usr/include/mysql -I/usr/include/mysql/mysql]
  --include       [-I/usr/include/mysql -I/usr/include/mysql/mysql]
  --libs          [-L/usr/lib64/ -lmariadb -lz -ldl -lm -lpthread -lssl -lcrypto]
  --libs_r        [-L/usr/lib64/ -lmariadb -lz -ldl -lm -lpthread -lssl -lcrypto]
  --libs_sys      [-lz -ldl -lm -lpthread -lssl -lcrypto]
  --version       [10.3.6]
  --cc_version    [3.0.8]
  --socket        [/var/lib/mysql/mysql.sock]
  --port          [3306]
  --plugindir     [/usr/lib64/mariadb/plugin]
  --tlsinfo       [OpenSSL 1.1.1]
```
As reported by @harmdewit, a symlink in the application working directory  made it works again.

**Comment by koenpunt** [comment]
Any update here on a proper fix, so not symlinking?

**Comment by defsprite** [comment]
I found a slightly less intrusive quickfix by setting `MARIADB_PLUGIN_DIR` to `/usr/lib/mariadb/plugin` as an environment variable.

**Comment by harmdewit** [comment]
Encountered this issue again, however it seems like it is fixed since Alpine 3.11. The mariadb-connector-c package was updated, on release mentions a fix about the plugin directory: https://mariadb.com/kb/en/mariadb-connector-c-313-release-notes/

I guess this wasn't a bug in the mariadb package, so closing the issue.
